lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1131,7 +1131,48 @@ TypeExpr *PreCheckExpression::simplifyTypeExpr(Expr *E) {
                                         ResultTypeRepr);
     return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
   }
-
+  
+  // Fold 'P & Q' into a composition type
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(E)) {
+    auto Operator = Lexer::getTokenAtLocation(TC.Context.SourceMgr,
```

Comment:
Rather than looking at the token, it would be better to retrieve the expression that describes the function, which will be an UnresolvedDeclRefExpr or DeclRefExpr... either way, you'll have direct access to the operator name without having to go back to the original source code.